### ISSUE TITLE
[qt5-base:x64-windows] build failure

### ISSUE THREAD

**Issue by BenKey** [microsoft/vcpkg#13566]
# Host Environment

- OS: Windows 10 Pro 64-bit Version 2004 build 19041.508 with Windows Feature Experience Pack 120.2212.31.0
- Compiler: Microsoft Visual Studio Community 2019 Version 16.7.3

# To Reproduce

Run the following command.

`./vcpkg install qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`

# Failure logs

## From Console

```
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:141 (message):
    Command failed: D:/vcpkg/downloads/tools/jom/jom-1.1.3/jom.exe /J 9
    Working Directory: D:/vcpkg/buildtrees/qt5-base/x64-windows-dbg
    See logs for more information:
      D:\vcpkg\buildtrees\qt5-base\build-x64-windows-dbg-out.log
      D:\vcpkg\buildtrees\qt5-base\build-x64-windows-dbg-err.log
```

## From build-x64-windows-dbg-out.log

```
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_IDXGIFactory2
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3D11DeviceContext1
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3DUserDefinedAnnotation
```

# Log files

[build-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/5232811/build-x64-windows-dbg-out.log)
[build-x64-windows-dbg-err.log](https://github.com/microsoft/vcpkg/files/5232814/build-x64-windows-dbg-err.log)

**Comment by JackBoosY** [comment]
```
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_IDXGIFactory2
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3D11DeviceContext1
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3DUserDefinedAnnotation
..\..\lib\Qt5Guid.dll : fatal error LNK1120: 3 unresolved externals
```
Should be linked with `uuid`.
cc @Neumann-A

**Comment by Neumann-A** [comment]
@JackBoosY: I still don't know where this error is coming from. So this needs a way to actually reproduce the error.

**Comment by BenKey** [comment]
I was able to reproduce the issue with the following command.

`./vcpkg install qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`

If you cannot reproduce the issue with that command, then it is possible that there are other packages that contribute to the issue. Try the following command instead.

`./vcpkg install directxsdk directxtk12 dx qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`

**Comment by BenKey** [comment]
@Neumann-A said.
> @JackBoosY: I still don't know where this error is coming from. So this needs a way to actually reproduce the error.

I provided information on how to reproduce this in the original bug report. I simply ran the following command and saw the errors.

`./vcpkg install qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`

When I saw your message I provided additional information 11 days ago. I will repeat my response from 11 days ago and provide a bit more information as explanation.

It is possible that there are other packages that contribute to the issue. Try the following command instead.

`./vcpkg install directxsdk directxtk12 dx qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`

The problem is that the original command was not run in isolation. I already had many other packages installed when I ran the command I mentioned in the original bug report. I can only guess why we are seeing differences in behavior. Since the errors reference DirectX and there are DirectX related packages that were installed before I ran that first command I assume the second one will suffice to reproduce the error.

I provided most of these details 11 days ago and have not seen a response.

Are you still unable to reproduce the issue?

I ask because I have not seen any progress on this issue in 11 days. I saw your comment and the requires:repro tag. I provided addition information, and nothing.

It seems to me that 11 days is more than enough time for someone to say "that worked" or "nope, still cannot reproduce it."

**Comment by JackBoosY** [comment]
@BenKey Sorry for late, I'll try to reproduce this issue later.

**Comment by JackBoosY** [comment]
After run command `./vcpkg install directxsdk directxtk12 dx qt5 qwt qt5[doc] qt5[speech] qt5-winextras --triplet x64-windows`, this issue was reproduced:
```
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_IDXGIFactory2
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3D11DeviceContext1
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3DUserDefinedAnnotation
```

**Comment by BenKey** [comment]
@JackBister Thanks.

**Comment by JackBoosY** [comment]
The symblos `IID_IDXGIFactory2` / `IID_ID3D11DeviceContext1` / `IID_ID3DUserDefinedAnnotation` are in dxguid.lib, and it has been declared to be used in _SOURCE/src/gui/rih/rhi.pri_:
```
win32 {
    HEADERS += \
        rhi/qrhid3d11_p.h \
        rhi/qrhid3d11_p_p.h
    SOURCES += \
        rhi/qrhid3d11.cpp

    LIBS += -ld3d11 -ldxgi -ldxguid
}
```
but not in the link command:
```
link /NOLOGO /DYNAMICBASE /NXCOMPAT /DEBUG /DLL /SUBSYSTEM:WINDOWS /VERSION:5.15 /MANIFEST:embed /OUT:..\..\lib\Qt5Guid.dll @C:\Users\vzay\AppData\Local\Temp\Qt5Guid.dll.1564.134140.jom
```

@Neumann-A do you know why it happened?

**Comment by Neumann-A** [comment]
> LIBS += -ld3d11 -ldxgi -ldxguid

should probably be

`LIBS += -ld3d11 -ldxgi -ldxguid -luuid`

**Comment by Neumann-A** [comment]
@JackBoosY
> but not in the link command:

You have to look into the response file `@C:\Users\vzay\AppData\Local\Temp\Qt5Guid.dll.1564.134140.jom` to get full cmd line call, switch to nmake ortake a peek into the makefile.

**Comment by JackBoosY** [comment]
I met this issue again.

After some investigation, I think this issue will repro if port `directxsdk` is installed first.
This should be because the dx library installed in vcpkg is **inconsistent** with the dx library in the Windows SDK, and vcpkg first uses the installed library in vcpkg.
It has been confirmed that these symbols do not exist in the `dxguid.lib` installed by vcpkg.
@BillyONeal  Is this a bug of directxsdk?

**Comment by JackBoosY** [comment]
@walbourn Do you know the difference between the dxsdk downloaded from the our official website and the dxsdk in the Windows 10 SDK?

**Comment by BillyONeal** [comment]
I observe directxsdk is explicitly deprecated in favor of the Windows SDK one and that port prints this message:

> message(WARNING "Build ${PORT} is deprecated, untested in CI, and requires the use of the DirectSetup legacy REDIST solution. See https://aka.ms/dxsdk for more information.")

**Comment by JackBoosY** [comment]
Obviously, qt5 did not process the code according to the version of dxsdk.
@Neumann-A Is there any qt document mentioned this issue?

**Comment by walbourn** [comment]
**TL;DR:** The only good reason to keep using the legacy DirectX SDK is for Windows XP. This scenario is not supported by the modern Windows SDK (Windows 8 or later) and requires the Windows 7.1A SDK.

@JackBoosY - The legacy DirectX SDK was deprecated back in 2011 as a product and lives on as an 'end-of-life' release with numerous known issues. See [Microsoft Docs](https://docs.microsoft.com/en-us/windows/win32/directx-sdk--august-2009-).  Due to the SHA-1 deprecation on Microsoft Downloads, there were impacts early in 2021 on both the legacy DirectX SDK and the legacy DirectX End-User Runtime downloads. See [this blog post](https://aka.ms/dxsdk). That's why I made changes to the community created **dxsdk** vcpkg port, as well as to provide features to support the guidance in [this blog post](https://walbourn.github.io/the-zombie-directx-sdk/).

The big problem with the legacy DirectX SDK is that it contains conflicting filenames with the modern Windows SDK. If you get the include order wrong, you'll end up with the old version of the header. That's almost certainly what's happening with ``dxguid.lib``. The copy of ``dxguid.lib`` in the legacy DirectX SDK does not contain ``IID_IDXGIFactory2`` or ``IID_IDXGIFactory2``.

A number of technologies were transferred to the Windows 8 OS and Window 8 SDK and live on in the modern Windows SDK: 

* Direct3D 11 headers/libs, Direct3D 10 headers/libs, Direct3D 9 headers/libs, DirectSound headers/libs, DirectInput headers/libs, DirectDraw headers.

> See [this blog post](https://walbourn.github.io/whither-directdraw/) for a note on why DirectDraw doesn't have a library in the Windows SDK.

* D3DCompile and D3DCSX APIs are supported in the Windows SDK, and the _46 and _47 DLLs are available to ship side-by-side for Windows 7 support. Those DLLs do NOT support Windows XP. See [this blog post](https://walbourn.github.io/hlsl-fxc-and-d3dcompile/). The ``FXC.EXE`` HLSL compiler tool was added to the Windows SDK as well.

* The "xnamath" library from the legacy DirectX SDK was replaced the DirectXMath library which is in the Window SDK. There's also a **directxmath** vcpkg port which provides the latest version of it. See [Microsoft Docs](https://docs.microsoft.com/en-us/windows/win32/dxmath/pg-xnamath-migration).

* The XInput headers in the Windows SDK use XInput 1.4 if building for Windows 8 or later, or XInput "9.1.0" if building below that. XInput "9.1.0" is a basic form of XInput that has been in the OS since Windows Vista. XInput 1.3 also has known security bugs, so the strong recommendation is to just stick with the basic gamepad functionality unless you are targeting Windows 8 or later. See [this blog post](https://walbourn.github.io/xinput-and-windows-8/).

* The XAudio2 headers in the Windows SDK require XAudio 2.8 or XAudio 2.9 which only support Windows 8 or later. We now have a [Microsoft.XAudio2.Redist](https://www.nuget.org/packages/Microsoft.XAudio2.Redist/) NuGet which offers XAudio 2.9 on Windows 7 or later so you don't need the legacy DXSDK at all for audio. This is exposed with the **xaudio2redist** vcpkg port. See [Microsoft Docs](https://aka.ms/XAudio2Redist).

* D3DX9, D3DX10, and D3DX11 are *not* in the Windows SDK and are officially deprecated. These were the primary reason people still install the legacy DirectX SDK for the past decade. We now have a [Microsoft.DXSDK.D3DX](https://www.nuget.org/packages/Microsoft.DXSDK.D3DX) NuGet package which both avoids the need to install the legacy DirectX SDK, *and* includes the required DLLs with a side-by-side license so you can also avoid the legacy DirectX End-User Runtime REDIST. See [this blog post](https://walbourn.github.io/legacy-d3dx-on-nuget/). This is exposed by the **dxsdk-d3dx** vcpkg port.

> The strong recommendation is to stop using legacy D3DX entirely. I have put a lot of work into open source replacements for this functionality over the years. That's the **directxtk**, **directxtex**, **directxmesh**, and **uvatlas** vcpkg ports expose. See [this blog post](https://aka.ms/Kfsdiu).

* The ``DXERR.LIB`` is not in the Windows SDK. See [this blog post](https://walbourn.github.io/wheres-dxerr-lib/) for a replacement.

* Lots of other legacy DirectX SDK stuff was left out of this transition. I've spent a lot of personal time finding new homes for most of them. See these blog posts: [DirectX SDK Tools Catalog](https://walbourn.github.io/directx-sdk-tools-catalog/), [DirectX SDK Samples Catalog](https://walbourn.github.io/directx-sdk-samples-catalog/).

**Comment by JackBoosY** [comment]
@walbourn Thank you for your detailed explanation, I will report this issue to qt.

**Comment by JackBoosY** [comment]
Upstream issue: https://bugreports.qt.io/browse/QTBUG-99050.

**Comment by Neumann-A** [comment]
This wont be fixed. 5.15.2 was the last open source release. So unless vcpkg switches to the kde branch for qt5 or adds the patch itself nothing will happen.

**Comment by JackBoosY** [comment]
@Neumann-A  I'm afraid the latest qt version also have this issue.

**Comment by mcDandy** [comment]
reproducable on windows 11 with wisual studio 17.1 and ninja 1.10.2

**Comment by JackBoosY** [comment]
Close this PR until upstream fixes this.
